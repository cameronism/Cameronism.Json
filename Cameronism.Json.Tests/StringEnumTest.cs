using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xunit;

namespace Cameronism.Json.Tests
{
	public unsafe class StringEnumTest
	{
		[Fact]
		public void DayOfWeekEnum()
		{
			var buffer = new byte[1024];
			int stringStart;
			int freeStart;

			fixed (byte* ptr = buffer)
			{
				var lookup = StringEnum.GenerateSequentialLookup(typeof(DayOfWeek), ptr, buffer.Length, out freeStart);

				Assert.True(lookup.HasValue, "lookup must have been generated");
				stringStart = (int)(lookup.Value.StringStart - ptr);
				
				//var chars = new char[64];
				//foreach (DayOfWeek dow in Enum.GetValues(typeof(DayOfWeek)))
				//{
				//	byte* str;
				//	int len;
				//	lookup.Value.FindSequential((ulong)dow, out str, out len);
				//	Assert.True(str != null, "string must not be null");
				//	var dowString = dow.ToString();
				//	Assert.Equal(dowString.Length, len);

				//	string actualString;
				//	fixed (char* ch = chars)
				//	{
				//		var strLen = Encoding.UTF8.GetChars(str, len, ch, chars.Length);
				//		actualString = new string(ch, 0, strLen);
				//	}
				//	Assert.Equal(dowString, actualString);
				//}
			}

			var sb = new StringBuilder();
			sb.AppendLine("## Lookups");
			Util.Hex.Dump(sb, buffer.Take(stringStart));
			sb.AppendLine();
			sb.AppendLine("## Strings");
			Util.Hex.Dump(sb, buffer.Skip(stringStart).Take(freeStart - stringStart));

			ApprovalTests.Approvals.Verify(sb.ToString());
		}
	}
}
