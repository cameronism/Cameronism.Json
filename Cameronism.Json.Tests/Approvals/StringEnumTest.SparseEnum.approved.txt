# Sorted Cameronism.Json.Tests.StringEnumTest+Sparse

GetLookupType: Sorted
Values count: 32
Max value: 2147483648

## Lookups


0000   01 00 00 00 00 00 02 00   02 00 00 00 02 00 02 00   ........ ........
0010   04 00 00 00 04 00 02 00   08 00 00 00 06 00 02 00   ........ ........
0020   10 00 00 00 08 00 02 00   20 00 00 00 0a 00 02 00   ........  .......
0030   40 00 00 00 0c 00 02 00   80 00 00 00 0e 00 02 00   @....... ........
0040   00 01 00 00 10 00 02 00   00 02 00 00 12 00 03 00   ........ ........
0050   00 04 00 00 15 00 03 00   00 08 00 00 18 00 03 00   ........ ........
0060   00 10 00 00 1b 00 03 00   00 20 00 00 1e 00 03 00   ........ . ......
0070   00 40 00 00 21 00 03 00   00 80 00 00 24 00 03 00   .@..!... ....$...
0080   00 00 01 00 27 00 03 00   00 00 02 00 2a 00 03 00   ....'... ....*...
0090   00 00 04 00 2d 00 03 00   00 00 08 00 30 00 03 00   ....-... ....0...
00a0   00 00 10 00 33 00 03 00   00 00 20 00 36 00 03 00   ....3... .. .6...
00b0   00 00 40 00 39 00 03 00   00 00 80 00 3c 00 03 00   ..@.9... ....<...
00c0   00 00 00 01 3f 00 03 00   00 00 00 02 42 00 03 00   ....?... ....B...
00d0   00 00 00 04 45 00 03 00   00 00 00 08 48 00 03 00   ....E... ....H...
00e0   00 00 00 10 4b 00 03 00   00 00 00 20 4e 00 03 00   ....K... ... N...
00f0   00 00 00 40 51 00 03 00   00 00 00 80 54 00 03 00   ...@Q... ....T...

## Strings


0000   76 31 76 32 76 33 76 34   76 35 76 36 76 37 76 38   v1v2v3v4 v5v6v7v8
0010   76 39 76 31 30 76 31 31   76 31 32 76 31 33 76 31   v9v10v11 v12v13v1
0020   34 76 31 35 76 31 36 76   31 37 76 31 38 76 31 39   4v15v16v 17v18v19
0030   76 32 30 76 32 31 76 32   32 76 32 33 76 32 34 76   v20v21v2 2v23v24v
0040   32 35 76 32 36 76 32 37   76 32 38 76 32 39 76 33   25v26v27 v28v29v3
0050   30 76 33 31 76 33 32                                0v31v32